File: ios/MenCare/Core/Network/Dto/CrmDto.swift
import Foundation

// Rubrica clienti — docs/API.md §Clienti (CRM).

struct ClientDto: Decodable {
    let id: String
    let firstName: String
    let lastName: String
    let phone: String
    let email: String?
    let customerSince: LocalDate
    let visitCount: Int
    /// Il totale speso lo vede solo il titolare: allo staff il campo non arriva
    /// proprio, e qui vale 0.
    let lifetimeSpendCents: Int64?
    let noShowCount: Int
    let lastVisit: LocalDate?
    let preferredServiceIds: [String]
    let preferredOperatorId: String?
    let marketingOptIn: Bool

    func toDomain() -> ClientRecord {
        ClientRecord(
            id: id,
            firstName: firstName,
            lastName: lastName,
            phone: phone,
            email: email ?? "",
            customerSince: customerSince,
            // Visite, ultima visita e spesa sono statistiche del server: le app
            // non le ricalcolano piu' scorrendo lo storico.
            visitCount: visitCount,
            lifetimeSpendCents: lifetimeSpendCents ?? 0,
            noShowCount: noShowCount,
            lastVisit: lastVisit,
            preferredServiceIds: preferredServiceIds,
            preferredOperatorId: preferredOperatorId,
            marketingOptIn: marketingOptIn
        )
    }
}

struct ClientsPageDto: Decodable {
    let clients: [ClientDto]
    let nextCursor: String?
}

/// `GET /crm/clients/:id` — scheda piu' storico gia' pronto, cosi' la schermata
/// non deve incrociare due chiamate.
struct ClientDetailDto: Decodable {
    struct HistoryEntry: Decodable {
        let id: String
        let operatorId: String
        let date: LocalDate
        let time: LocalTime
        let durationMinutes: Int
        let status: String
        let channel: String
        let serviceIds: [String]
        let serviceNames: [String]
        let totalPriceCents: Int64?

        func toDomain(clientId: String) -> Appointment {
            Appointment(
                id: id,
                clientId: clientId,
                operatorId: operatorId,
                serviceIds: serviceIds,
                start: date.atTime(time),
                durationMinutes: durationMinutes,
                totalPriceCents: totalPriceCents ?? 0,
                status: WireEnum.appointmentStatus(status),
                channel: WireEnum.channel(channel)
            )
        }
    }

    /// Abitudini calcolate dal server sulle visite completate.
    struct Insights: Decodable {
        let favoriteOperatorId: String?
        let averageDaysBetweenVisits: Int?
    }

    let client: ClientDto
    let insights: Insights?
    let appointments: [HistoryEntry]
}

struct CreateClientRequestDto: Encodable {
    let firstName: String
    let lastName: String
    let phone: String
    let email: String?
}

File: android/core/network/src/main/kotlin/com/devora/mencare/core/network/dto/CrmDto.kt
package com.devora.mencare.core.network.dto

import kotlinx.serialization.Serializable

/**
 * Rubrica clienti: `/v1/crm/…`.
 *
 * `lifetimeSpendCents` manca nelle risposte al ruolo STAFF — non è nascosto
 * dalla schermata, non viene proprio selezionato — quindi qui è opzionale.
 */

@Serializable
data class ClientDto(
    val id: String,
    val firstName: String,
    val lastName: String,
    val phone: String,
    val email: String? = null,
    val customerSince: String,
    val visitCount: Int = 0,
    val lifetimeSpendCents: Long? = null,
    val noShowCount: Int = 0,
    val lastVisit: String? = null,
    val preferredServiceIds: List<String> = emptyList(),
    val preferredOperatorId: String? = null,
    val marketingOptIn: Boolean = true,
)

/** Pagina della rubrica: `nextCursor` è `null` quando non c'è altro da leggere. */
@Serializable
data class ClientPageDto(
    val clients: List<ClientDto> = emptyList(),
    val nextCursor: String? = null,
)

/** Riga dello storico del cliente: più ricca dell'appuntamento, porta i nomi dei servizi. */
@Serializable
data class ClientHistoryDto(
    val id: String,
    val operatorId: String,
    val date: String,
    val time: String,
    val startsAt: String? = null,
    val durationMinutes: Int,
    val status: String,
    val channel: String = "APP",
    val serviceIds: List<String> = emptyList(),
    val serviceNames: List<String> = emptyList(),
    val totalPriceCents: Long? = null,
)

/** Abitudini calcolate dal server sulle visite completate. */
@Serializable
data class ClientInsightsDto(
    val favoriteOperatorId: String? = null,
    val averageDaysBetweenVisits: Int? = null,
)

@Serializable
data class ClientDetailDto(
    val client: ClientDto,
    val insights: ClientInsightsDto? = null,
    val appointments: List<ClientHistoryDto> = emptyList(),
)

@Serializable
data class CreateClientBody(
    val firstName: String,
    val lastName: String,
    val phone: String,
    val email: String? = null,
)
